rotkehlchen/exchanges/coinbaseprime.py

```diff
@@ -0,0 +1,458 @@
+        try:
+            response = self.session.get(url=uri, params=params)
+        except requests.RequestException as e:
+            raise RemoteError(f'Coinbase Prime API request failed due to {e}') from e
+
+        try:
+            data = response.json()
+        except requests.JSONDecodeError as e:
+            raise RemoteError(f'Coinbase Prime returned invalid json {response.text}') from e
+
+        return data
+
+    @overload
+    def _query_paginated_endpoint(
+            self,
+            query_params: dict[str, Any],
+            portfolio_id: str,
+            method: Literal['transactions'],
+            decoding_logic: Callable[[dict[str, Any]], AssetMovement | None],
+    ) -> list[AssetMovement]:
+        ...
+
+    @overload
+    def _query_paginated_endpoint(
+            self,
+            query_params: dict[str, Any],
+            portfolio_id: str,
+            method: Literal['orders'],
+            decoding_logic: Callable[[dict[str, Any]], Trade | None],
+    ) -> list[Trade]:
+        ...
+
+    def _query_paginated_endpoint(
+            self,
+            query_params: dict[str, Any],
+            portfolio_id: str,
+            method: Literal['orders', 'transactions'],
+            decoding_logic: Callable[[dict[str, Any]], AssetMovement | None] | Callable[[dict[str, Any]], Trade | None],  # noqa: E501
+    ) -> list[Trade] | list[AssetMovement]:
+        """Abstraction to consume all the events in the selected queries.
+        It uses the `decoding_logic` to process the different events and returns a list
+        that depends on the signature of it.
+
+        This function may raise:
+        - RemoteError
+        """
+        result = []
+        while True:
+            response = self._api_query(
+                module='portfolios',
+                path=f'{portfolio_id}/{method}',
+                params=query_params,
+            )
+
+            events = response[method]
+            for raw_event in events:
+                try:
+                    if (event := decoding_logic(raw_event)) is None:
+                        log.warning(
+                            f'Wont process event {raw_event} from coinbase prime. Skipping',
+                        )
+                    else:
+                        result.append(event)
+                except DeserializationError as e:
+                    self.msg_aggregator.add_error(
+                        f'Failed to process coinbase prime trade due to {e}. Skipping entry...',
+                    )
+
+            if (
+                response['pagination']['has_next'] is True and
+                (new_cursor := response['pagination'].get('next_cursor', '')) != ''
+            ):
+                query_params['cursor'] = new_cursor
+            else:
+                break
+
+        return result  # type: ignore  # mypy doesn't detect that the return type here is defined by the function used
```

Comment:
It can't since `self._api_query` returns a dict, a key of whose is different. You are not showing this difference with an overload at that method.

Reply:
The return of `_api_query` doesn't affect to the variable result. Notice that I'm using `event := decoding_logic(raw_event)` and then `result.append(event)` is the signature of `decoding_logic` what should affect the signature of result